From 98ce6624cd2563cfdb5da3b2949d5e1e03867034 Mon Sep 17 00:00:00 2001
From: Krishna Kondaka <kkondaka@fb.com>
Date: Tue, 30 Jul 2019 11:32:42 -0700
Subject: [PATCH] Enforce a limit on value bytes size

Summary:
Enforce a limit on value bytes size

This is a fix for CVE-2019-11923

Reviewed By: stuclar

Differential Revision: D16471999

fbshipit-source-id: 9d614da8534e20935b1561c613bb81defd7d470a
---
 mcrouter/lib/network/McAsciiParser.h  | 3 +++
 mcrouter/lib/network/McAsciiParser.rl | 4 ++++
 2 files changed, 7 insertions(+)

diff --git a/mcrouter/lib/network/McAsciiParser.h b/mcrouter/lib/network/McAsciiParser.h
index f24a7aa9..9d2f010d 100644
--- a/mcrouter/lib/network/McAsciiParser.h
+++ b/mcrouter/lib/network/McAsciiParser.h
@@ -81,6 +81,9 @@ class McAsciiParserBase {
       const char* posStart,
       const char* posEnd);
 
+  // limit the value size.
+  static constexpr uint32_t maxValueBytes = 1 * 1024 * 1024 * 1024; // 1GB
+
   std::string currentErrorDescription_;
 
   uint64_t currentUInt_{0};
diff --git a/mcrouter/lib/network/McAsciiParser.rl b/mcrouter/lib/network/McAsciiParser.rl
index 36e02085..3ff07814 100644
--- a/mcrouter/lib/network/McAsciiParser.rl
+++ b/mcrouter/lib/network/McAsciiParser.rl
@@ -137,6 +137,10 @@ exptime_req = negative? uint %{
 
 value_bytes = uint %{
   remainingIOBufLength_ = static_cast<size_t>(currentUInt_);
+  // Enforce maximum on value size obtained from parser                       
+  if (remainingIOBufLength_ > maxValueBytes) {
+    remainingIOBufLength_ = maxValueBytes;                                    
+  }
 };
 
 cas_id = uint %{
